[WebGPU] WGPUBindGroupEntry::offset is not used while creating the bind group

mwyrzykowski opened the pull request:
#### 6a81017aa27aca988157447ba83d5d545f48b169
<pre>
[WebGPU] WGPUBindGroupEntry::offset is not used while creating the bind group
<a href="https://bugs.webkit.org/show_bug.cgi?id=249834">https://bugs.webkit.org/show_bug.cgi?id=249834</a>
&lt;radar://103606837&gt;

Reviewed by Myles C. Maxfield.

Use data offset member while creating the bind group.

This allows amongst other things sharing one buffer across
multiple bind groups where each bind group has a unique result.

Attached sample modeled after <a href="https://austin-eng.com/webgpu-samples/samples/twoCubes">https://austin-eng.com/webgpu-samples/samples/twoCubes</a>

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::sizeOfEntries):
(WebGPU::Device::createBindGroup):
Correct aliasing of buffers where multiple bind groups share
the same bind group layout.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::isValid const):
(WebGPU::BindGroupLayout::vertexArgumentBuffer const): Deleted.
(WebGPU::BindGroupLayout::fragmentArgumentBuffer const): Deleted.
(WebGPU::BindGroupLayout::computeArgumentBuffer const): Deleted.
Bind group layout shouldn&apos;t contain MTLBuffer instances since one
bind group layout can be shared amongst multiple bind groups.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::stagesForBinding const):
(WebGPU::BindGroupLayout::~BindGroupLayout): Deleted.
Can default this now that the union is removed.

(WebGPU::BindGroupLayout::stageForBinding const): Deleted.

* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/two-cubes.html: Added.
Sample creates two bind groups from the same buffer with different offsets.

Canonical link: <a href="https://commits.webkit.org/258513@main">https://commits.webkit.org/258513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d8eb62b570508c4aba37abd472c3ee65c5158be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110731 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11569 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1527 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108548 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92042 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35326 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90699 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78325 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4219 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24961 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4282 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1430 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10368 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44450 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5837 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6044 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -77,34 +110,39 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+                        return BindGroup::createInvalid(*this);
+
+                    ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
+                    *(float**)(argumentBufferContents[currentStage] + offset) = (float*)((char*)buffer.gpuAddress + entry.offset);
```
Is it really correct to assume the buffer is `float*`, or is the entry just an address sized value?

mwyrzykowski replied:
The assumption is the entry is an address sized value in the generated metal source, either a reference or a pointer type

djg commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -77,34 +110,39 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
         }
 
-        return BindGroup::create(vertexArgumentBuffer, fragmentArgumentBuffer, computeArgumentBuffer, WTFMove(resources), *this);
+        return BindGroup::create(argumentBuffer[0], argumentBuffer[1], argumentBuffer[2], WTFMove(resources), *this);
```
I won't be opposed to something like `argumentBuffer[vertexStage]` instead of `argumentBuffer[0]`.

mwyrzykowski replied:
I will make this change during the PR for https://bugs.webkit.org/show_bug.cgi?id=250123 which I am working on now

litherum reviewed (approved):
rs=me because this is part of https://github.com/WebKit/WebKit/pull/8153

mwyrzykowski committed:
[WebGPU] WGPUBindGroupEntry::offset is not used while creating the bind group
https://bugs.webkit.org/show_bug.cgi?id=249834
<radar://103606837>

Reviewed by Myles C. Maxfield.

Use data offset member while creating the bind group.

This allows amongst other things sharing one buffer across
multiple bind groups where each bind group has a unique result.

Attached sample modeled after https://austin-eng.com/webgpu-samples/samples/twoCubes

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::sizeOfEntries):
(WebGPU::Device::createBindGroup):
Correct aliasing of buffers where multiple bind groups share
the same bind group layout.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::isValid const):
(WebGPU::BindGroupLayout::vertexArgumentBuffer const): Deleted.
(WebGPU::BindGroupLayout::fragmentArgumentBuffer const): Deleted.
(WebGPU::BindGroupLayout::computeArgumentBuffer const): Deleted.
Bind group layout shouldn't contain MTLBuffer instances since one
bind group layout can be shared amongst multiple bind groups.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::stagesForBinding const):
(WebGPU::BindGroupLayout::~BindGroupLayout): Deleted.
Can default this now that the union is removed.

(WebGPU::BindGroupLayout::stageForBinding const): Deleted.

* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/two-cubes.html: Added.
Sample creates two bind groups from the same buffer with different offsets.

Canonical link: https://commits.webkit.org/258513@main
